modules/agent_engine/main.tf

```diff
@@ -0,0 +1,208 @@
+
+          content {
+            name  = env.key
+            value = env.value
+          }
+        }
+
+        dynamic "secret_env" {
+          for_each = var.secret_environment_variables
+
+          content {
+            name = secret_env.key
+
+            secret_ref {
+              secret  = secret_env.value.secret_id
+              version = secret_env.value.version
+            }
+          }
+        }
+      }
+    }
+
+    package_spec {
+      python_version           = var.python_version
+      dependency_files_gcs_uri = "gs://${local.bucket_name}/${google_storage_bucket_object.dependencies.name}"
+      requirements_gcs_uri     = "gs://${local.bucket_name}/${google_storage_bucket_object.requirements.name}"
+      pickle_object_gcs_uri = (
+        var.generate_pickle
+        ? "gs://${local.bucket_name}/${google_storage_bucket_object.pickle_from_src[0].name}"
+        : "gs://${local.bucket_name}/${google_storage_bucket_object.pickle[0].name}"
+      )
+    }
+  }
+}
+
+# Eventual consistency issue.
+# AE doesn't retry (yet) the deployment if bindings are still not active.
+resource "time_sleep" "wait_5_minutes" {
+  create_duration = "5m"
+
+  depends_on = [
+    google_project_iam_member.default
+  ]
+}
+
+resource "google_storage_bucket" "default" {
+  count                       = var.bucket_config.create ? 1 : 0
+  name                        = coalesce(var.bucket_config.name, var.name)
+  project                     = local.project_id
+  location                    = local.location
+  uniform_bucket_level_access = var.bucket_config.uniform_bucket_level_access
+  force_destroy               = !var.bucket_config.deletion_protection
+}
+
+resource "null_resource" "default" {
+  count = var.generate_pickle ? 1 : 0
+
+  provisioner "local-exec" {
+    command = join(" ", [
+      "python",
+      "./tools/serialize_agent.py",
+      "${var.source_files.path}/${var.source_files.pickle_src}",
+      "--output-file ${var.source_files.path}/${var.source_files.pickle_out}",
+      "--variable-name ${var.source_files.pickle_src_var_name}"
+    ])
+  }
+}
+
+resource "google_storage_bucket_object" "dependencies" {
+  name   = "${filesha1("${var.source_files.path}/${var.source_files.dependencies}")}.tar.gz"
+  bucket = local.bucket_name
+  source = "${var.source_files.path}/${var.source_files.dependencies}"
+}
+
+resource "google_storage_bucket_object" "pickle_from_src" {
+  count  = var.generate_pickle ? 1 : 0
+  name   = "${filesha1("${var.source_files.path}/${var.source_files.pickle_src}")}.pkl"
+  bucket = local.bucket_name
+  source = "${var.source_files.path}/${var.source_files.pickle_out}"
+
+  depends_on = [
+    null_resource.default
+  ]
+}
+
+resource "google_storage_bucket_object" "pickle" {
+  count  = var.generate_pickle ? 0 : 1
+  name   = "${filesha1("${var.source_files.path}/${var.source_files.pickle_out}")}.pkl"
+  bucket = local.bucket_name
+  source = "${var.source_files.path}/${var.source_files.pickle_out}"
```

Comment:
Please add also `source_md5hash` - to calculate it properly you may need to add `local_file` resource that depends on your `null_resource`.

This is the only way to make sure, that if you update this, it won't crash the provider.

Reply:
so you're saying I can keep the normal name (i.e. pickle.pkl) and use detect_md5hash-1 = true ...right?

Reply:
I mean the warning that's there, and [source_md5](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket_object#source_md5hash-1) field.

No, I mean creating:
```hcl
resource "local_file" pickle {
 path = "to/pickle"
}
...

resource "google_storage_bucket_object" "..." {
  ...
  source_md5hash = local_file.pickle.content_md5
  ...
}
```